src/dataflow/src/source/s3.rs

```diff
@@ -836,6 +810,66 @@ async fn download_object(
     }
 }
 
+async fn read_object_chunked<R>(
+    source_id: &str,
+    object_length: Option<i64>,
+    reader: &mut R,
+    tx: &Sender<Result<InternalMessage, S3Error>>,
+) -> (DownloadStatus, Option<DownloadMetricUpdate>)
+where
+    R: Unpin + AsyncRead,
+{
+    let chunk_size = cmp::min(
```

Comment:
`tokio_util::io::ReadStream` will allow you to express the logic of this function with just stream combinators. I took at stab at it here https://github.com/MaterializeInc/materialize/commit/99de61c04b45a9701a568356056a5f7a89c68b8b

Reply:
Nice. It does clean things up a bit, although I had to use a loop because the error branch of `try_fold` doesn't give you access to the state.